### ISSUE TITLE
itch app sandbox and LÖVE 11.1

### ISSUE THREAD

**Issue by fluffy-critter** [itchio/itch#1850]
I updated two of my games, http://fluffy.itch.io/refactor and http://fluffy.itch.io/cat-catcher, to use LÖVE 11.1. Ever since doing so, the macOS itch.io app cannot launch them; they just start up and immediately segfault.

Other games of mine running LÖVE 0.10.2 still launch just fine, and if I disable itch's sandbox mode these games will launch as well.

Presumably this is actually a bug in LÖVE, but I'd like to know if there's something that could be done differently on the itch side, or what actual sandboxing API itch is using so that I can furnish this information to the LÖVE devs.

Thanks!

**Comment by fasterthanlime** [comment]
Hey fluffy, thanks for the report!

We do have a minimal amount of docs on the sandbox - it works differently on each platform, here's the macOS docs:

  * https://itch.io/docs/itch/using/sandbox/macos.html

I'll try running the game myself on macOS eventually, but if you could have a quick look at Console.app messages like the docs say, I might be able to figure it out without launching the ol' VM :)

**Comment by fasterthanlime** [comment]
I'm looking at this right now and I *think* it's caused by app translocation. It's hard to tell because not every app breaks with it in the sandbox, but I have a few good leads. I'll keep investigating.

**Comment by fasterthanlime** [comment]
Here's my theory so far:

  * We whitelist the original app bundle's location
  * The sandboxing system creates another .app bundle in a temporary location
  * One of these gets translocated (possibly both?)
  * love11/newer SDL2 apps try to open themselves for reading - and segfault if they can't
  * They can't because they have access to the original path, but not the translocated path

Possible fixes:

  * Figure out translocated path, and adjust sandbox to give access to that
  * Remove 'quarantine' xattr from the original or shim bundle or both, idk

This is all very obscure, thanks Apple :(

**Comment by retropipes** [comment]
Removing the quarantine xattr is super-easy to do for a single file: `/usr/bin/xattr -d com.apple.quarantine <file>` ... and for a folder, recursively: `/usr/bin/xattr -d -r com.apple.quarantine <folder>` - much like we now do with hdiutil to handle disk images.

**Comment by fasterthanlime** [comment]
Yep, I've had a chance to learn about the quarantine xattr when debugging this: https://github.com/itchio/itch-setup/issues/3

..but I'm still not sure this is what's happening in this case. Further investigation is needed, which is annoying because I don't have a physical mac available at the moment.

One way you could help test this @wrldwzrd89 is:

  * Start itch with `~/Applications/itch.app/Contents/MacOS/itch` in a terminal, so you can see real-time output
  * Make sure the sandbox is enabled in your settings
  * Install either of the games @fluffy-critter mentioned
  * Make sure they crash on launch
  * See if the installed .app bundle has the quarantine xattr
  * If it does, remove the xattr and try to launch it again.

Now that I'm writing this, I'm not sure why it would have that xattr, because we're not downloading it via a browser, we're writing it to disk directly, so I think the issue is just that it isn't signed, so it gets translocated anyway :(

Maybe figuring out the translocated app prefix and giving read-only access to that folder would be enough? I'm still very confused by this issue.

---

Note: If you want to take a look at the shim bundle created, you can `export INVESTIGATE_SANDBOX=1` before launching, see the code here: https://github.com/itchio/smaug/blob/5c88b6a507f35961e3f1eded67289241d66d799d/runner/sandboxexec_darwin.go#L19 - it will hang forever instead of launching, but you'll be able to look at the bundle, even modify it, and open it manually.
